### PR TITLE
assignment thirteen advanced language features

### PR DESCRIPTION
**Pull request by Sruthi-Subaraja**
In progress

ASSIGNMENT 13 - ADVANCED LANGUAGE FEATURES
TASK 1 - Understanding and Implementing Events and Delegates in C#.
TASK 2 - Understanding the Use of Dynamic and Var Keywords and Their Differences.
TASK 3 - Implementing Anonymous Methods.

### CODE REVIEW

File: src/AdvancedLanguageFeatures/Program.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                    return (T)Convert.ChangeType(input, typeof(T));
+                }
+                else
+                {
+                    continue;
+                }
+            }
+
+            Console.WriteLine("More than three wrong inputs. Please try again");
+            return (T)Convert.ChangeType(-1, typeof(T));
```
Comment (Tharun-Sundar) [inline src/AdvancedLanguageFeatures/Program.cs:109]: What happens when you return (T)(-1) instead?

File: src/AdvancedLanguageFeatures/Program.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        public static T ValidateInput<T>(string messageToPrint, string regexToCompareWith)
+        {
+            for (int i = 0; i < 3; i++)
+            {
+                string input = GetUserInput(messageToPrint);
+                if (IsInputValid(input, regexToCompareWith))
+                {
+                    return (T)Convert.ChangeType(input, typeof(T));
+                }
+                else
```
Comment (Tharun-Sundar) [inline src/AdvancedLanguageFeatures/Program.cs:102]: As you are returning with the `if`, the else is not needed.

File: src/AdvancedLanguageFeatures/Program.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        {
+            var inputOne = 1;
+
+            // inputOne = "one"; The var keyword does not allow the type of the variable to be changed after it is initially declared.
+            dynamic inputTwo = 2;
+            inputTwo = "two";
+        }
+
+        /// <summary>
+        /// Task three using anonymous function for comparison.
```
Comment (Tharun-Sundar) [inline src/AdvancedLanguageFeatures/Program.cs:53]:
```suggestion
        /// Task three: Using anonymous function for comparison.
```

File: src/AdvancedLanguageFeatures/Program.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        public static void TaskOne()
+        {
+            Notifier notifier = new ();
+            notifier.OnAction += PrintMessage;
+            string messageToPrint = ValidateInput<string>("Enter any message to Print", "(.*?)");
+            notifier.OnOnAction(messageToPrint);
+        }
+
+        /// <summary>
+        /// Task two difference between var and dynamic
```
Comment (Tharun-Sundar) [inline src/AdvancedLanguageFeatures/Program.cs:41]:
```suggestion
        /// Task two: Difference between var and dynamic
```

File: src/AdvancedLanguageFeatures/Program.cs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        }
+
+        /// <summary>
+        /// Task one invoking an event
+        /// </summary>
+        public static void TaskOne()
+        {
+            Notifier notifier = new ();
+            notifier.OnAction += PrintMessage;
+            string messageToPrint = ValidateInput<string>("Enter any message to Print", "(.*?)");
```
Comment (Tharun-Sundar) [inline src/AdvancedLanguageFeatures/Program.cs:36]:
```suggestion
            string messageToPrint = ValidateInput<string>("Enter any message to print", "(.*?)");
```

File: src/AdvancedLanguageFeatures/Program.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            TaskTwo();
+            int[] array = new int[] { 5, 4, 3, 8, 2, 9, 12, 13 };
+            TaskThree(array);
+
+            // TaskFour();
+            // TaskFive();
+        }
+
+        /// <summary>
+        /// Task one invoking an event
```
Comment (Tharun-Sundar) [inline src/AdvancedLanguageFeatures/Program.cs:30]:
```suggestion
        /// Task one: Invoking an event
```

**Review by Tharun-Sundar — CHANGES_REQUESTED** [review]

**Review by Tharun-Sundar — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -24,14 +24,19 @@ TASK 3 - Using garbage collection and understanding it's impact on performance.
...
+TASK 4 - Debugging tools and tips
+TASK 5 - Troubleshooting and issue resolution
 ### ASSIGNMENT 10 -ERROR HANDLING
 TASK 1 - Understanding and using try/catch/finally blocks.
 TASK 2 - Catching and throwing different types of exceptions.
 TASK 3 - Defining and using custom exception classes.
 TASK 4 - Handling global unhandled exceptions.
 TASK 5 - Understanding stack trace.
+### ASSIGNMENT 13 - ADVANCED LANGUAGE FEATURES
+TASK 1 - Understanding and Implementing Events and Delegates in C#.
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply README.md:39]: Changed.

File: README.md [hunk 8]
```diff
@@ -24,14 +24,19 @@ TASK 3 - Using garbage collection and understanding it's impact on performance.
...
+TASK 2 - Understanding the Use of Dynamic and Var Keywords and Their Differences.
+TASK 3 - Implementing Anonymous Methods.
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply README.md:41]: Done

File: src/AdvancedLanguageFeatures/Program.cs (hunk 2 above)
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:102]: Okay

File: src/AdvancedLanguageFeatures/Program.cs (hunk 4 above)
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:41]: Okay

File: src/AdvancedLanguageFeatures/Program.cs (hunk 6 above)
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:30]: Okay

File: src/AdvancedLanguageFeatures/Program.cs (hunk 3 above)
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:53]: Okay

File: README.md [hunk 9]
```diff
@@ -24,14 +24,19 @@ TASK 3 - Using garbage collection and understanding it's impact on performance.
...
+TASK 5 - Troubleshooting and issue resolution
 ### ASSIGNMENT 10 -ERROR HANDLING
 TASK 1 - Understanding and using try/catch/finally blocks.
 TASK 2 - Catching and throwing different types of exceptions.
 TASK 3 - Defining and using custom exception classes.
 TASK 4 - Handling global unhandled exceptions.
 TASK 5 - Understanding stack trace.
+### ASSIGNMENT 13 - ADVANCED LANGUAGE FEATURES
+TASK 1 - Understanding and Implementing Events and Delegates in C#.
+TASK 2 - Understanding the Use of Dynamic and Var Keywords and Their Differences.
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply README.md:40]: Done

File: src/AdvancedLanguageFeatures/Program.cs (hunk 5 above)
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:36]: Done

File: src/AdvancedLanguageFeatures/Program.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        /// <summary>
+        /// Gets valid input from user.
+        /// </summary>
+        /// <param name="messageToPrint">The message printed for the user</param>
+        /// <param name="regexToCompareWith">Regex to compare with</param>
+        /// <typeparam name="T">String or integer</typeparam>
+        /// <returns>valid string or integer value</returns>
+        public static T ValidateInput<T>(string messageToPrint, string regexToCompareWith)
+        {
+            for (int i = 0; i < 3; i++)
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:95]: Okay Tharun

File: src/AdvancedLanguageFeatures/Program.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        public static void PrintMessage(string message)
+        {
+            Console.WriteLine(message);
+        }
+
+        /// <summary>
+        /// Gets valid input from user.
+        /// </summary>
+        /// <param name="messageToPrint">The message printed for the user</param>
+        /// <param name="regexToCompareWith">Regex to compare with</param>
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:90]: Done

File: src/AdvancedLanguageFeatures/Program.cs [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            return input;
+        }
+
+        /// <summary>
+        /// Checks if the input is valid.
+        /// </summary>
+        /// <param name="input">The input that is checked</param>
+        /// <param name="regexToCompareWith">Regex the input will be compared with</param>
+        /// <returns>Returns true if input is valid</returns>
+        public static bool IsInputValid(string input,  string regexToCompareWith)
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:130]: Done

File: src/AdvancedLanguageFeatures/Program.cs [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        /// <summary>
+        /// Print user message and return input
+        /// </summary>
+        /// <param name="messageToPrint">Message to prompt the user</param>
+        /// <returns>User input</returns>
+        public static string GetUserInput(string messageToPrint)
+        {
+            Console.WriteLine(messageToPrint);
+            string? input = Console.ReadLine();
+            return input;
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:121]: Okay

File: src/AdvancedLanguageFeatures/Program.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        }
+
+        /// <summary>
+        /// Checks if the input is valid.
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:125]: Done

File: src/AdvancedLanguageFeatures/Program.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+        /// <summary>
+        /// Print user message and return input
+        /// </summary>
+        /// <param name="messageToPrint">Message to prompt the user</param>
+        /// <returns>User input</returns>
+        public static string GetUserInput(string messageToPrint)
+        {
+            Console.WriteLine(messageToPrint);
+            string? input = Console.ReadLine();
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:120]: Okay

File: src/AdvancedLanguageFeatures/Program.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        /// <param name="message">The message to print to the user</param>
+        public static void PrintMessage(string message)
+        {
+            Console.WriteLine(message);
+        }
+
+        /// <summary>
+        /// Gets valid input from user.
+        /// </summary>
+        /// <param name="messageToPrint">The message printed for the user</param>
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:89]: Okay

File: src/AdvancedLanguageFeatures/Program.cs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+        /// <param name="regexToCompareWith">Regex to compare with</param>
+        /// <typeparam name="T">String or integer</typeparam>
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:91]: Done

File: src/AdvancedLanguageFeatures/Program.cs (hunk 1 above)
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:109]: We can't convert int type to T type

File: tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs [hunk 18]
```diff
@@ -0,0 +1,144 @@
+namespace TaskManagerClassTests
+{
+    using AdvancedLanguageFeatures;
+    public class TaskManagerClassTests
+    {
```
Comment (Tharun-Sundar) [inline tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs:5]: Test names need to be updated

**Review by Tharun-Sundar — CHANGES_REQUESTED** [review]

File: src/AdvancedLanguageFeatures/Notifier.cs [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace AdvancedLanguageFeatures
+{
+    /// <summary>
+    /// Notifier class to perform actions when an event arises
+    /// </summary>
+    public class Notifier
+    {
+        /// <summary>
+        /// Notifies the user with message
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Notifier.cs]: Changed it to a more appropriate summary.

File: src/AdvancedLanguageFeatures/Notifier.cs [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+        /// </summary>
+        /// <param name="message">The message printed to the user</param>
+        public delegate void Notify(string message);
+
+        /// <summary>
+        /// Event of Notify delegate type
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Notifier.cs]: Changed it.

File: src/AdvancedLanguageFeatures/TaskManager.cs [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+                case Rectangle rectangle:
+                    rectangle.PrintDetails();
+                    break;
+                case Triangle triangle:
+                    triangle.PrintDetails();
+                    break;
+                case Circle circle:
+                    circle.PrintDetails();
+                    break;
+                case null:
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/TaskManager.cs:144]: It is not needed.

File: src/AdvancedLanguageFeatures/TaskManager.cs [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Display a record using deconstruction.
+        /// </summary>
+        /// <param name="book">Record object to print.</param>
+        public void Display(Book book)
+        {
+            var (title, author, isbn) = book;
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/TaskManager.cs:157]: Yes, it is the requirement to use deconstruction to print the book details.

File: src/AdvancedLanguageFeatures/Shape.cs [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+        public Shape(string? color)
+        {
+            this.Color = color;
+        }
+
+        /// <summary>
+        /// Gets or sets the color of the shape.
+        /// </summary>
+        /// <value>Color of the shape.</value>
+        public string? Color { get; set; }
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Shape.cs:25]: Okay

File: src/AdvancedLanguageFeatures/TaskManager.cs [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+        }
+
+        /// <summary>
+        /// Task four: Understanding and using lambda expressions and statements.
+        /// </summary>
+        /// <param name="number">List of integer</param>
+        /// <returns>List of integer with only squared odd numbers</returns>
+        public List<int> TaskFour(List<int> number)
+        {
+            number = number.Where(x => x % 2 != 0).ToList();
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/TaskManager.cs:71]: Okay done

File: src/AdvancedLanguageFeatures/Program.cs [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        /// <summary>
+        /// Checks if the input matches the regex provided.
+        /// </summary>
+        /// <param name="input">The input that is checked</param>
+        /// <param name="regexToCompare">Regex the input will be compared with</param>
+        /// <returns>Returns true if input is valid otherwise false</returns>
+        public static bool IsInputValid(string input, string regexToCompare)
+        {
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:127]: Okay

File: src/AdvancedLanguageFeatures/Program.cs [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+
+        /// <summary>
+        /// Prints array.
+        /// </summary>
+        /// <param name="array">Array to be printed</param>
+        public static void PrintArray(int[] array)
+        {
+            for (int i = 0; i < array.Length; i++)
+            {
+                Console.Write(array[i] + ",");
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Program.cs:53]: Okay

File: src/AdvancedLanguageFeatures/Notifier.cs [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+        /// </summary>
+        public event Notify? OnAction;
+
+        /// <summary>
+        /// To invoke the event
+        /// </summary>
+        /// <param name="message">The message to print to the user</param>
+        public virtual void OnOnAction(string message)
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Notifier.cs]: It is not needed

File: src/AdvancedLanguageFeatures/Circle.cs [hunk 28]
```diff
@@ -0,0 +1,51 @@
+﻿// <copyright file="Circle.cs" company="PlaceholderCompany">
+// Copyright (c) PlaceholderCompany. All rights reserved.
+// </copyright>
+
+namespace AdvancedLanguageFeatures
+{
+    /// <summary>
+    /// Circle class with properties radius and calculates area with it.
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AdvancedLanguageFeatures/Circle.cs:8]: Okay

File: tests/unit/AdvancedLanguageFeatures/NotifierClassTests/NotifierClassTest.cs [hunk 29]
```diff
@@ -0,0 +1,19 @@
+namespace NotifierClassTests
+{
+    using AdvancedLanguageFeatures;
+    public class NotifierClassTest
+    {
+        [Fact]
+        public void NoEvent_InvokeOnActionEvent_ExecuteTheMethodsSubscribed()
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply tests/unit/AdvancedLanguageFeatures/NotifierClassTests/NotifierClassTest.cs:7]: Done

File: tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs (hunk 18 above)
Comment (Sruthi-Subaraja) [reply tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs:5]: Done

File: tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+            int[] unsortedArray = {5, 34, 2, 12, 24};
+            int[] expectedArray = { 2, 5, 12, 24, 34 };
+
+            int[] actualArray = taskManager.RunTaskThree(unsortedArray);
+
+            Assert.Equal(expectedArray, actualArray);
+        }
+
+        [Fact]
+        public void ListWithPositiveAndNegativeNumbers_RunTaskFour_ReturnSquareOfTheOddNumbersOnly()
```
Comment (Tharun-Sundar) [inline tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs:19]: The test given part says there will be a list with positive and negative numbers. But the input `listofNumber` only has positive numbers.

**Review by Tharun-Sundar — APPROVED** [review]

File: tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs [hunk 31]
```diff
@@ -0,0 +1,148 @@
...
+            List<int> listOfNumber = new List<int>() { 2, 5, 6, 3, 23, 12};
+            List<int> expectedList = new List<int>() { 25, 9, 529 };
+
+            List<int> actualList = taskManager.RunTaskFour(listOfNumber);
+
+            Assert.Equal(expectedList, actualList);
+        }
+
+        [Fact]
+        public void TwoProducts_ProductNamesAreCompared_ReturnPositiveInteger()
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs:31]: Okay

File: tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs (hunk 30 above)
Comment (Sruthi-Subaraja) [reply tests/unit/AdvancedLanguageFeatures/TaskManagerClassTests/TaskManagerClassTests.cs:19]: The list will contain odd and even numbers. I made a mistake naming it positive and negative integers.

**Pull request merged**